212 tree for scopes

vukihub opened the pull request:
Implemented the node tree, save log configuration in the file, fixed couple of raised bugs.

aregtech committed:
Implemented saving config file. Commit for testing.

aregtech committed:
Enabled build and run unit tests, removed forcing build examples.

aregtech commented on framework/areg/base/private/FileBase.cpp:
```diff
@@ -104,7 +104,7 @@ inline int _readLine(const FileBase & file, ClassType & outValue)
             if ( readLength != 0 )
             {
-                readLength = NEString::removeChar<CharType>(static_cast<CharType>('\r'), buffer, static_cast<NEString::CharCount>(readLength), true);
+                // readLength = NEString::removeChar<CharType>(static_cast<CharType>('\r'), buffer, static_cast<NEString::CharCount>(readLength), true);
```
Remove comment

aregtech commented on framework/areg/trace/GETrace.h:
```diff
@@ -67,6 +67,37 @@
+    #define PRIO_FATAL          (static_cast<unsigned int>(NETrace::eLogPriority::PrioFatal)    | static_cast<unsigned int>(NETrace::eLogPriority::PrioScope))
+
+    //!< Unset logging scopes in priorities
+    #define PRIO_NOSCOPES(x)    ((x)& (~static_cast<unsigned int>(NETrace::eLogPriority::PrioScope)))
```
Add a space here `((x)& (`

aregtech commented on framework/areg/trace/GETrace.h:
```diff
@@ -67,6 +67,37 @@
+     * \brief   Change scope log priority during runtime.
+     *          Pass scope set in DEF_TRACE_SCOPE and priority value with OR operation specified in NETrace::eLogPriority.
+     **/
+    #define SCOPE_PRIORITY_CHANGE(scope, prio)          NETrace::scopePriorityChange(#scope, static_cast<unsigned int>(prio))
```
add a macro to get a priority

aregtech commented on framework/areg/trace/GETrace.h:
```diff
@@ -182,15 +218,51 @@
+     * \brief   Returns true if logging is enabled
+     **/
+    #define IS_LOG_ENABLED()                            ((3-2) > 0)
+
```
Format in line 322 and 326

aregtech commented on framework/areg/trace/NETrace.hpp:
```diff
@@ -86,6 +118,8 @@ namespace NETrace
      **/
     AREG_API NETrace::eLogPriority convFromString( const String& strPrio );
 
+    AREG_API String makePrioString( unsigned int priorities );
```
Document the method

aregtech commented on framework/areg/trace/NETrace.hpp:
```diff
@@ -914,16 +989,45 @@ inline const char* NETrace::getString(NETrace::eLogPriority prio)
         return "NETrace::PrioIgnoreLayout";
     case NETrace::eLogPriority::PrioAny:
         return "NETrace::PrioAny";
+    case NETrace::eLogPriority::PrioValid:
```
not all enums are present. check again

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
-            file.writeString( properties[ i ].makeString( ) );
+            newProperty.clearProperty( );
         }
+    }
```
At the end, check if `newProperty` has any data (can be invalid), add to the list.

aregtech commented on framework/areg/trace/private/NETrace.cpp:
```diff
@@ -172,7 +176,45 @@ AREG_API_IMPL NETrace::eLogPriority NETrace::convFromString( const String& strPr
+
+AREG_API_IMPL String NETrace::makePrioString( unsigned int priorities )
+{
+    String result;
```
Enable conversion if `AREG_LOGS` is set.

aregtech commented on framework/areg/trace/private/ScopeNodeBase.cpp:
```diff
@@ -0,0 +1,284 @@
+    else
+    {
+        return false;
+    }
```
Optimize
```cpp
return (mNodeType == other.mNodeType ? (mNodeName > other.mNodeName) : (mNodeType > other.mNodeType);
```

aregtech commented on framework/areg/trace/private/ScopeNodeBase.cpp:
```diff
@@ -0,0 +1,284 @@
+    else
+    {
+        return true;
+    }
```
Optimize
```cpp
return (mNodeType == other.mNodeType ? (mNodeName < other.mNodeName) : (mNodeType < other.mNodeType);
```

aregtech committed:
Update ScopeNodeBase.cpp

aregtech commented on framework/areg/trace/private/ScopeNodeBase.hpp:
```diff
@@ -0,0 +1,481 @@
+    /**
+     * \brief   The grouping flags to set bitwise.
+     **/
+    enum class eGroupping
```
```cpp
enum class eGroupping : unsigned char
```

aregtech commented on framework/areg/trace/private/ScopeNodes.cpp:
```diff
@@ -0,0 +1,429 @@
+
+    if ( child.isLeaf( ) )
+    {
+        std::pair<LeafList::LISTPOS, bool> atPos = mChildLeafs.addIfUnique( ScopeLeaf( child ), false );
```
`std::pair<LeafList::LISTPOS, bool>` can be `auto`, internal variable

aregtech commented on framework/areg/trace/private/ScopeNodes.cpp:
```diff
@@ -0,0 +1,429 @@
+    }
+    else if (child.isNode( ))
+    {
+        std::pair<NodeList::LISTPOS, bool> atPos = mChildNodes.addIfUnique( ScopeNode( child ), false );
```
`std::pair<LeafList::LISTPOS, bool>` can be `auto`, internal variable

aregtech commented on framework/areg/trace/private/ScopeNodes.cpp:
```diff
@@ -0,0 +1,429 @@
+    result += Process::getInstance( ).getAppName( );
+    result += NELogConfig::SYNTAX_OBJECT_SEPARATOR;
+
+    return result;
```
```cpp
String result;
return result.format("%s.%s.", NELogConfig::SYNTAX_CMD_LOG_SCOPE.data(), Process::getInstance( ).getAppName( ).getString());
```

aregtech commented on framework/areg/trace/private/ScopeNodes.cpp:
```diff
@@ -0,0 +1,429 @@
+    char buffer[ size ];
+    String::formatString( buffer, size, "%s* = %s", parent.getString(), prio.getString() );
+
+    return String( buffer );
```
```cpp
String result;
return result.format("%s* = %s"), parent.getString(), makePrioString( ).getString());
```

aregtech commented on framework/areg/trace/private/ScopeNodes.hpp:
```diff
@@ -0,0 +1,376 @@
+    /**
+     * \brief   Returns true if a node has no children nodes.
+     **/
+    virtual bool isEmpty( void ) const;
```
`override` at the end

aregtech commented on framework/areg/trace/private/ScopeNodes.hpp:
```diff
@@ -0,0 +1,376 @@
+     * \param   parentPath  The path name of the parent node. The nodes contain '_' at the end.
+     * \return  Returns the number of nodes that have been saved.
+     **/
+    virtual unsigned int saveNodeConfig( FileBase & file, const String & parentPath ) const;
```
`override` at the end

aregtech commented on framework/areg/trace/private/TraceProperty.cpp:
```diff
@@ -108,17 +108,27 @@ void TraceProperty::clearProperty( bool clearComment /* = true */ )
-    result += NELogConfig::SYNTAX_EQUAL;
-    result += static_cast<String>(mProperty.mValue.second);
-    result += NELogConfig::SYNTAX_LINEEND;
+    String result;
```
`String result(mComment)` at least the comment should be there.

aregtech commented on framework/areg/trace/private/TraceProperty.cpp:
```diff
@@ -108,17 +108,27 @@ void TraceProperty::clearProperty( bool clearComment /* = true */ )
+
+        constexpr int size{ 512 };
+        char buffer[ size ];
+        String::formatString( buffer, size, "%s = %s\n", mProperty.mValue.first.getKey( ).getString( ), static_cast<const char *>(mProperty.mValue.second) );
```
```cpp
String config;
result += config.format("%s = %s", mProperty.mValue.first.getKey( ).getString( ), static_cast<const char *>(mProperty.mValue.second));
```

aregtech commented on framework/areg/trace/private/TracePropertyKey.cpp:
```diff
@@ -106,21 +106,23 @@ void TracePropertyKey::setKey(const String & fullKey)
+        else
+        {
+            mConfigKey = NELogConfig::eConfigKey::KeyLocal;
+        }
```
```cpp
mConfigKey = (mModuleName.isEmpty( ) || (mModuleName == NELogConfig::SYNTAX_SCOPE_GROUP)) ? NELogConfig::eConfigKey::KeyGlobal : NELogConfig::eConfigKey::KeyLocal;
```

aregtech commented on tests/units/LogScopesTest.cpp:
```diff
@@ -0,0 +1,390 @@
+DEF_TRACE_SCOPE( areg_unit_tests_LogScopeTest_ScopePriorityGroupping_infoNode2_noScope );
+DEF_TRACE_SCOPE( areg_unit_tests_LogScopeTest_ScopePriorityGroupping_information );
+TEST( LogScopeTest, ScopePriorityGroupping )
+{
```
In this test would be good to save the scope priority and later compare the values when load config file.

aregtech reviewed (changes requested):
Please see the comments.

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
-    path = File::getFileFullPath( File::normalizePath( path ) );
-    File fileConfig( path, FileBase::FO_MODE_EXIST | FileBase::FO_MODE_READ | FileBase::FO_MODE_TEXT | FileBase::FO_MODE_SHARE_READ );
-    fileConfig.open( );
+    constexpr unsigned int mode{ FileBase::FO_MODE_WRITE | FileBase::FO_MODE_TEXT | FileBase::FO_MODE_CREATE | FileBase::FO_MODE_SHARE_READ | FileBase::FO_MODE_SHARE_WRITE };
```
The `FileBase::FO_MODE_SHARE_WRITE` can be removed. We don't want to share

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
-        file.moveToBegin( );
-        PropertyList properties( mProperties );
-        PropertyList scopes;
+        ASSERT( false );
```
remove assert

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
-        while ( file.readLine( line ) > 0 )
+    if ( file.isOpened( ) == false )
+    {
+        ASSERT( false );
```
remove assert

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
+
+    if ( file.canWrite( ) == false )
+    {
+        ASSERT( false );
```
remove assert

aregtech commented on framework/areg/trace/private/LogConfiguration.cpp:
```diff
@@ -93,68 +94,121 @@ bool LogConfiguration::loadConfig( FileBase & file )
+    // path = File::getFileFullPath( File::normalizePath( path ) );
+    File fileConfig( path, mode );
+    
+    VERIFY(fileConfig.open( ));
```
Not sure whether we need `VERIFY`

aregtech reviewed (changes requested):
Added more comments

aregtech committed:
Fixed requested changes. Fixed writing file in linux.

aregtech committed:
checking the failure on ci/cd

aregtech committed:
Update common.cmake

aregtech committed:
upload failed ut

trying to upload failed unit tests

aregtech committed:
another try to upload failed tests

another try to upload failed tests

aregtech committed:
fixed value

fixed the value

aregtech committed:
Update LogScopesTest.cpp

Set working directory

aregtech committed:
Update LogScopesTest.cpp

aregtech committed:
Fixing unit test for windows.

aregtech committed:
Experimenting with gtest

aregtech committed:
Fix unit test under windows

aregtech committed:
assertion too small, experiment

aregtech committed:
Experimenting to solve problem on github unit test run for windows.

aregtech committed:
Temporary enable only unit tests with shared areg

Disabled all builds, except unit tests with shared library for all platform until find the solution for windows.

aregtech committed:
Experimental file system unit tests.

aregtech committed:
Merge branch 'feature/212-tree-for-scopes' of https://github.com/aregtech/areg-sdk into feature/212-tree-for-scopes

aregtech committed:
Areg File test to read and write.

aregtech committed:
Added couple of more tests with areg

aregtech committed:
search the bug, narrow the build.

aregtech committed:
unit test of normalize path

aregtech committed:
Fixing linux build

aregtech committed:
Slightly increased the size of buffers.

aregtech committed:
Analyze log start and stop bug

aregtech committed:
Log bug

aregtech committed:
Fixing log start stop

aregtech committed:
Fixing paths

aregtech committed:
Fixed the unit test file name

aregtech committed:
Fixed linux build. Output test messages

aregtech committed:
testing local time conversion

aregtech committed:
TestLocalTimeWin32 is only for windows

aregtech committed:
Fixed timezone info.

aregtech committed:
add new unit test to check the buffer size

aregtech committed:
Fixed unit test bug. There is still one left to check the time.

aregtech reviewed (approved):
looks good
